File: SistemaGestionTransporte/src/gestorAplicaciones/entidades/Empleado.java
package gestorAplicaciones.entidades;

import java.util.ArrayList;

public class Empleado extends Usuario {
	
	//atributos
	private final static ArrayList<Empleado> empleados = new ArrayList<Empleado>();

	private boolean estatusActivo;
	private boolean disponible;
	private String pais;
	private String ciudadActual;
	
	//constructor
	public Empleado (String nombre, String clave, long id, String correo) {
		super(nombre, clave, id, correo);
	}

	public Empleado (String nombre, String clave, long id, String correo, boolean statusActivo, String pais, String ciudadActual){
		this(nombre, clave, id, correo);
		this.estatusActivo = statusActivo;
		this.pais = pais;
		this.ciudadActual = ciudadActual;
		Empleado.empleados.add(this);
	}
	
	//metodos getter and setter
	public static ArrayList<Empleado> getEmpleados() {
		return empleados;
	}


	public boolean isEstatusActivo() {
		return estatusActivo;
	}

	public void setEstatusActivo(boolean estatusActivo) {
		this.estatusActivo = estatusActivo;
	}

	public boolean isDisponible() {
		return disponible;
	}

	public void setDisponible(boolean disponible) {
		this.disponible = disponible;
	}

	public String getPais() {
		return pais;
	}

	public void setPais(String pais) {
		this.pais = pais;
	}

	public String getCiudadActual() {
		return ciudadActual;
	}

	public void setCiudadActual(String ciudadActual) {
		this.ciudadActual = ciudadActual;
	}

	//metodos

	public boolean elegirConductor(String ciudadActual){
		return this.ciudadActual.equals(ciudadActual) && this.estatusActivo && this.disponible;
	}

	public static Empleado seleccionarEmpleado(String origen) {
		//buscar un empleado disponible y que se encuentre en la ciudad de origen
		for(Empleado empleado : Empleado.getEmpleados()){
			if(empleado.elegirConductor(origen)) return empleado;
		}
		return null;
	}

	public String toString(){
		//sobreescritura del metodo toString para mostar los atributos del objeto.
		return "\nnombre\t" + this.getNombre()
				+ "\nid\t" + this.getID()
				+ "\ncorreo\t" + this.getCorreo()
				+ "\nActivo\t" + this.estatusActivo
				+ "\nPais\t" + this.pais
				+ "\n";
	}

	public double calcularPago(double costoPedido) {
		return costoPedido;
	}

	public static void crearEmpleado(String nombre, String clave, long id, String correo, String pais, String ciudadActual) {
		new Empleado(nombre, clave, id, correo, true, pais,ciudadActual);
	}

	public static boolean isNombreValido(String nombre){
        /*
        Esta funcion retorna false si el argumento que se ingresa no contione espacio y caracteres
        alfabetiscos o si ya es un atributo de nombre de un objeto de tipo usuario,
        de lo contrario retorna true.
         */
		if(nombre == null || !nombre.chars().allMatch(c -> c == ' ' || Character.isLetter(c))){
			System.out.println("El nombre solo debe tener caracteres alfanumericos.");
			return false;
		}
		for(Empleado empleado : Empleado.empleados) {
			if (empleado.comprobarNombre(nombre)) {
				System.out.println("nombre ya registrado.");
				return false;
			}
		}
		return true;

	}

	public static boolean isIDValido(String id){
        /*
        Esta funcion retorna false si el argumento que se ingresa no contione solo caracteres
        numericos o si ya es un atributo de ID de un objeto de tipo usuario,
        de lo contrario retorna true.
         */
		if(id == null || !id.chars().allMatch(Character::isDigit)){
			System.out.println("la identificacion debe contener solo caracteres numericos.");
			return false;
		}
		for(Empleado empleado : Empleado.empleados) {
			if (empleado.comprobarID(Long.parseLong(id))) {
				System.out.println("numero de identificacion ya registrado.");
				return false;
			}
		}
		return true;
	}

	public static boolean isCorreoValido(String correo){
        /*
        Esta funcion retorna false si el argumento que se ingresa contiene el caracter '@' o si
        solo contiene en la ultima posicion, de lo contrario retorna true.
         */
		if(!correo.contains("@") || correo.charAt(correo.length()-1) == '@'){
			System.out.println("correo no valido.");
			return false;
		}
		for(Empleado empleado : Empleado.empleados) {
			if (empleado.comprobarCorreo(correo)) {
				System.out.println("correo ya registrado.");
				return false;
			}
		}
		return true;
	}

}

File: python/SistemaGestionTransporte/src/gestorAplicaciones/entidades/empleado.py
from src.gestorAplicaciones.entidades.usuario import Usuario


class Empleado(Usuario):
    empleados = []

    def __init__(self, nombre, clave, id, correo, estatusActivo, pais, ciudadActual):
        super().__init__(nombre, id, correo, clave)
        self.estatusActivo = estatusActivo
        self.pais = pais
        self.ciudadActual = ciudadActual
        self.disponible = True
        Empleado.empleados.append(self)

    @classmethod
    def getEmpleados(cls):
        return cls.empleados

    def isEstatusActivo(self):
        return self.estatusActivo

    def setEstatusActivo(self, estatusActivo):
        self.estatusActivo = estatusActivo

    def isDisponible(self):
        return self.disponible

    def setDisponible(self, disponible):
        self.disponible = disponible

    def getPais(self):
        return self.pais

    def setPais(self, pais):
        self.pais = pais

    def getCiudadActual(self):
        return self.ciudadActual

    def setCiudadActual(self, ciudadActual):
        self.ciudadActual = ciudadActual

    def elegirConductor(self, ciudadActual):
        return self.ciudadActual == ciudadActual and self.estatusActivo and self.disponible

    @classmethod
    def seleccionarEmpleado(cls, origen):
        for empleado in cls.getEmpleados():
            if empleado.elegirConductor(origen):
                return empleado
        return None

    def __str__(self):
        return "\nnombre:\t" + self.get_nombre() + \
            "\nid:\t" + str(self.get_id()) + \
            "\ncorreo:\t" + self.get_correo() + \
            "\nPais:" + self.getPais() + \
            "\nCiudad Actual:\t" + self.ciudadActual + \
            "\nActivo:\t" + str(self.estatusActivo) + \
            "\nDisponible:\t" + str(self.disponible) + \
            "\nPais\t" + self.pais + "\n"

    def calcularPago(self, costoPedido):
        return costoPedido + costoPedido * 0.2

    @classmethod
    def crearEmpleado(cls, nombre, clave, id, correo, pais, ciudadActual):
        empleado = Empleado(nombre, clave, id, correo)
        empleado.estatusActivo = True
        empleado.pais = pais
        empleado.ciudadActual = ciudadActual
        empleado.disponible = True
        cls.getEmpleados().append(empleado)

    @staticmethod
    def isNombreValido(nombre):
        if nombre is None or not nombre.isalpha():
            return False
        for empleado in Empleado.getEmpleados():
            if empleado.comprobar_nombre(nombre):
                return False
        return True

    @staticmethod
    def isIDValido(id):
        if id is None or not id.isdigit():
            print("La identificación debe contener solo caracteres numéricos.")
            return False
        for empleado in Empleado.getEmpleados():
            if empleado.comprobar_id(int(id)):
                print("Número de identificación ya registrado.")
                return False
        return True

    @staticmethod
    def isCorreoValido(correo):
        if "@" not in correo or correo[-1] == "@":
            print("Correo no válido.")
            return False
